### PR TITLE
Decode input secrets used in template stringData

### LINKED ISSUE #445: Secrets used as input resources are double base64 encoded when used in stringData
**What steps did you take:**

1. Author a new SecretTemplate that uses stringData to build a multiline secret
2. Specify inputResources that are of kind Secret
3. Reference the inputResources in the stringData field
4. Apply the SecretTemplate

Example:
```
apiVersion: secretgen.carvel.dev/v1alpha1
kind: SecretTemplate
metadata:
  name: my-templated-secret
spec:
  inputResources:
  - name: my-input-secret
    ref:
      apiVersion: v1
      kind: Secret
      name: a-password-secret
  template:
    stringData:
      config: |
        #This is an example
        someKey = $(.my-input-secret.data.password)
```

**What happened:**
The values from the password secret were retrieved (base64 encoded) and added to the secret, the resulting string in the stringData field was then base64 encoded again. Meaning when my application decodes the templated secret, it doesn't contain the password, but rather the base64 contents of the original secrets. This makes the templated secret unusable.

**What did you expect:**
I expected the values to be base64 decoded so that my application could use them.

**Anything else you would like to add:**

**Environment:**

- secretgen-controller version: v0.14.8
- Kubernetes version: v1.24.15

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by carpenterm**
If an input resource is of kind Secret, and that is used in an expression in the stringData field of a SecretTemplate, then base64 decode it when evaluating the stringData field.

Fixes #445: Secrets used as input resources are double base64 encoded when used in stringData

### CODE REVIEW

File: pkg/generator/secret_template_test.go [hunk 1]
```diff
@@ -89,6 +89,62 @@ func Test_SecretTemplate(t *testing.T) {
...
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            "secretTemplate",
+					Namespace:       "test",
+					ResourceVersion: "1",
+					OwnerReferences: []metav1.OwnerReference{
+						secretTemplateOwnerRef("secretTemplate"),
+					},
+				},
+				Data: map[string][]byte{
```
Comment (neil-hickey) [inline pkg/generator/secret_template_test.go:139]: 139-142 should be covered by the previous unit-test.

File: pkg/generator/secret_template_test.go [hunk 2]
```diff
@@ -89,6 +89,62 @@ func Test_SecretTemplate(t *testing.T) {
...
+					InputResources: []sg2v1alpha1.InputResource{{
+						Name: "creds",
+						Ref: sg2v1alpha1.InputResourceRef{
+							APIVersion: "v1",
+							Kind:       "Secret",
+							Name:       "existingSecret",
+						},
+					}},
+					JSONPathTemplate: &sg2v1alpha1.JSONPathTemplate{
+						Data: map[string]string{
```
Comment (neil-hickey) [inline pkg/generator/secret_template_test.go:109]: 109-112 should be covered by the previous unit-test. This test is only for StringData

File: pkg/generator/secret_template_reconciler.go [hunk 3]
```diff
@@ -283,7 +284,12 @@ func evaluateTemplate(template *sg2v1alpha1.JSONPathTemplate, values map[string]
 	}
 
 	// Template Secret StringData
-	stringData, err := evaluate(template.StringData, values)
+	decodedValues, err := decodeSecrets(values)
```
Comment (neil-hickey) [inline pkg/generator/secret_template_reconciler.go:287]: This is a pre-processing step which ends up having to iterate all the values even if you never specify a SecretData field. We also don't have any test for the "else" case, i.e.

```
if obj.GetKind() == "Secret" { } else { // no tests }
```



Is there a way to fold this into our existing `evaluate` function or a new function `evaluateStringData` which could delegate to the existing evaluateBytes or similar for most of the work.

**Comment by devthejo** [comment]
I need this too, can I do something to help with this ?
@carpenterm are you still on the case ?

**Comment by carpenterm** [comment]
Sorry but I am not working on this anymore, by the time the review came in I'd found another way using Kyverno.

**Pull request closed without merging**